velox/exec/prefixsort/benchmarks/PrefixSortAlgorithmBenchmark.cpp

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <folly/Benchmark.h>
+#include <folly/init/Init.h>
+#include "velox/buffer/Buffer.h"
+#include "velox/exec/prefixsort/PrefixSortAlgorithm.h"
+#include "velox/exec/prefixsort/PrefixSortEncoder.h"
+
+DEFINE_int32(sort_data_seed, 1, "random test data generate seed.");
+
+using namespace facebook::velox;
+using namespace facebook::velox::exec;
+
+namespace {
+class PrefixSortAlgorithmBenchmark {
+ public:
+  void seed(int32_t seed) {
+    rng_.seed(seed);
+  }
+
+  void runQuickSort(std::vector<long> vec) {
+    char* start = (char*)vec.data();
+    uint32_t entrySize = sizeof(long);
+    auto swapBuffer = AlignedBuffer::allocate<char>(entrySize, pool_.get());
+    auto context =
+        prefixsort::PrefixSortContext(entrySize, swapBuffer->asMutable<char>());
+    context.quickSort(
+        start, start + entrySize * vec.size(), [&](char* a, char* b) {
+          return memcmp(a, b, sizeof(long));
```

Comment:
We probably should use a prefix length not known to compiler.  Maybe keep it in a gflag and default to 8.

Reply:
>Maybe keep it in a gflag and default to 8
I test that ,  nothing different,  if think sizeof(long) is as a #define for compiler. 

something addition
I plan to use auto-vectorized tech when use the sort function , the code will like:

`
if (entrySize = 8) {
    context.quickSort(
        start, start + entrySize * vec.size(), [&](char* a, char* b) {
          return memcmp(a, b,  8);
}
if (entrySize = 9) {
    context.quickSort(
        start, start + entrySize * vec.size(), [&](char* a, char* b) {
          return memcmp(a, b,  9);
}
`
I have do some benchmark test , the performance is as same as I do this here.

Reply:
This can be different from compiler to compiler.  It's safer to keep it unknown to compiler to match the intended usage.

Reply:
Oh so you plan to specialize on some of these lengths or adding a simd::memcmp?  Then we can leave it as it is.

Reply:
> Also if you plan to specialize the prefix lengths, would it make sense to put it as a template parameter to the context so compiler will see it throughout the sorting process not just the compare function?

Yes,  'throughout the sorting process '  will avoid add if satement in compare method, this helps improve performance.

Reply:
`swap` could also be faster if entrySize_ is known to compiler

Reply:
> `swap` could also be faster if entrySize_ is known to compiler

I test use use constant value (directly use 8), the performance is same as we use simd::memcpy, that is 'entrySize_ is known to compiler' helps unroll loop , and  simd::memcpy could also get that.

Reply:
> would it make sense to put it as a template parameter to the context 

But compare method is a custom function , we need to do some code to handle keys cannot normalized, is there some tech to pass PrefixSortContext<entrySize> `s entrySize to compare method ?

Reply:
**For swap mthod :**

use constant memcpy
```
  FOLLY_ALWAYS_INLINE void swap(
      const detail::PrefixSortIterator& lhs,
      const detail::PrefixSortIterator& rhs) const {
    memcpy(swapBuffer_, *lhs, 8);
    memcpy(*lhs, *rhs, 8);
    memcpy(*rhs, swapBuffer_, 8);
  }
```

PrefixSort_algorithm_10k                                    1.10ms    909.85
PrefixSort_algorithm_100k                                  13.33ms     75.00
PrefixSort_algorithm_1000k                                157.54ms      6.35
PrefixSort_algorithm_10000k                                  1.90s   526.83m

```
  FOLLY_ALWAYS_INLINE void swap(
      const detail::PrefixSortIterator& lhs,
      const detail::PrefixSortIterator& rhs) const {
    simd::memcpy(swapBuffer_, *lhs, entrySize_);
    simd::memcpy(*lhs, *rhs, entrySize_);
    simd::memcpy(*rhs, swapBuffer_, entrySize_);
  }
```

PrefixSort_algorithm_10k                                    1.17ms    852.10
PrefixSort_algorithm_100k                                  14.29ms     69.96
PrefixSort_algorithm_1000k                                171.34ms      5.84
PrefixSort_algorithm_10000k                                  2.05s   487.97m

We can see there is nearly not different between the two method

@Yuhta  So could we leave PrefixSortConext as it is to reduce code complex ?

Reply:
I would say there is still some gap though.  How do we foresee the distribution of prefix lengths in this case?  If it is within a few fixed sizes we should defined it as compile time template.  If we need to support arbitrary lengths then leave it as runtime variable.

Reply:
> I would say there is still some gap though

'some gap ' you metioned here is memcpy or memcmp? I think, for memcpy, since we could use simd::memcpy we do not need to concern the arbitrary lengths case.

> How do we foresee the distribution of prefix lengths in this case? If it is within a few fixed sizes we should defined it as compile time template. If we need to support arbitrary lengths then leave it as runtime variable.

For memcmp + arbitrary lengths, there will be a `length dispacth`, duckDB implement dispacth the logic in memcpy itself, see:

https://github.com/duckdb/duckdb/blob/main/src/include/duckdb/common/fast_mem.hpp

```
	case 63:
		return MemcmpFixed<63>(str1, str2);
	case 64:
		return MemcmpFixed<64>(str1, str2);
	default:
		return memcmp(str1, str2, size);
```

I plan to implement the dispach logic outside custm-compare method to avoid add branch statment in hot code.


@Yuhta

Reply:
I mean there is still a gap in memcpy with constant size vs simd::memcpy.  If we already plan to dispatch on prefix length, we can just leverage the memcpy in this case.

Reply:
> just leverage the memcpy in this case

1. The code wil be a little complex for PrefixRunner(old name PrefixContext) to handle fixed-size / non-fixed-size, the function signature will be different. Maybe I am wrong about this, any similar example to define a class like this is helpful.
2. simd vs fixed-size, I prefer simd, because it make code simple, could we assume that we will finnaly got a simd::cmp method in future and can remove  'dispatch entrySize' code in memcmp case ?   I can provide a version for AVX now , but need more effort to handle other case.

Reply:
I see so dispatch is temporary.  In that case yes we can find a faster memcmp to replace this one, and you probably should put the dispatch in a place that is easy to replace with a simd::memcmp.

Another thing I have in mind is what if we treat the prefix not as `char*` but as `uint64_t`?  Can we make sure the prefix is alway a multiple of 8?  Then we don't need to do the `bswap` to encode and then `bswap` back for a faster comparison.

Reply:
>  Can we make sure the prefix is always a multiple of 8
No 
for example : order by col_short, col_tinyint，col_boolean ,  the min unit is 1 byte.